fe/fe-core/src/main/java/com/starrocks/catalog/SchemaInfo.java

```diff
@@ -0,0 +1,270 @@
+    public short getShortKeyColumnCount() {
+        return shortKeyColumnCount;
+    }
+
+    public KeysType getKeysType() {
+        return keysType;
+    }
+
+    public TStorageType getStorageType() {
+        return storageType;
+    }
+
+    public int getVersion() {
+        return version;
+    }
+
+    public List<Column> getColumns() {
+        return columns;
+    }
+
+    public List<Integer> getSortKeyIndexes() {
+        return sortKeyIndexes;
+    }
+
+    public List<Integer> getSortKeyUniqueIds() {
+        return sortKeyUniqueIds;
+    }
+
+    public List<Index> getIndexes() {
+        return indexes;
+    }
+
+    public Set<String> getBloomFilterColumnNames() {
+        return bloomFilterColumnNames;
+    }
+
+    public double getBloomFilterFpp() {
+        return bloomFilterFpp;
+    }
+
+    public TTabletSchema toTabletSchema() {
+        TTabletSchema tSchema = new TTabletSchema();
+        tSchema.setShort_key_column_count(shortKeyColumnCount);
+        tSchema.setKeys_type(keysType.toThrift());
+        tSchema.setStorage_type(storageType);
+        tSchema.setId(id);
+        tSchema.setSchema_version(version);
+        tSchema.setSchema_hash(schemaHash);
+        tSchema.setIs_in_memory(false/*unused now*/);
+
+        List<TColumn> tColumns = new ArrayList<TColumn>();
+        for (Column column : columns) {
+            TColumn tColumn = column.toThrift();
+            // is bloom filter column
+            if (bloomFilterColumnNames != null && bloomFilterColumnNames.contains(column.getName())) {
+                tColumn.setIs_bloom_filter_column(true);
+            }
+            // when doing schema change, some modified column has a prefix in name.
+            // this prefix is only used in FE, not visible to BE, so we should remove this prefix.
+            if (column.getName().startsWith(SchemaChangeHandler.SHADOW_NAME_PRFIX)) {
+                tColumn.setColumn_name(column.getName().substring(SchemaChangeHandler.SHADOW_NAME_PRFIX.length()));
+            }
+            if (column.getName().startsWith(SchemaChangeHandler.SHADOW_NAME_PRFIX_V1)) {
+                tColumn.setColumn_name(column.getName().substring(SchemaChangeHandler.SHADOW_NAME_PRFIX_V1.length()));
+            }
+            tColumns.add(tColumn);
+        }
+        tSchema.setColumns(tColumns);
+        tSchema.setSort_key_idxes(sortKeyIndexes);
+        tSchema.setSort_key_unique_ids(sortKeyUniqueIds);
+
+        if (CollectionUtils.isNotEmpty(indexes)) {
+            List<TOlapTableIndex> tIndexes = new ArrayList<>();
+            for (Index index : indexes) {
+                tIndexes.add(index.toThrift());
+            }
+            tSchema.setIndexes(tIndexes);
+        }
+
+        if (bloomFilterColumnNames != null) {
+            tSchema.setBloom_filter_fpp(bloomFilterFpp);
+        }
+        return tSchema;
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static class Builder {
```

Comment:
can this code be generated by protobuf? so just define a protobuf and generating the corresponding Builder code. The `TTabletSchema` class just needs to implement a protobuf to `TTabletSchema` convert method.

Reply:
The jprotobuf generated java code will not use the builder pattern.

Reply:
not necessary to be jprotobuf, just regular protobuf for auto code generation.

Reply:
It seems that protobufs defined in StarRocks are compiled with jprotobuf by default, I may not be quite sure of the way you mean. Also, this class will be used in the PR I'm submitting later, and it needs to support being serialized by gson (so its members are marked with SeralizedName). java classes generated from protobuf can also be serialized by gson, but it needs to be done with an additional byte array conversion, and maybe that will make things more complicated?

Reply:
ok